src/lib/chart/dendrogram.typ

```diff
@@ -0,0 +1,312 @@
+///                           These keys are used as argument to the
+///                           rows `.at(..)` function.
+/// - mode (string): Chart mode:
+///                  - `"vertical"` -- Vertically displayed dendrogram
+/// - size (array): Chart size as width and height tuple in canvas units;
+///                 height can be set to `auto`.
+/// - line-style (style,function): Style or function (idx => style) to use for
+///                               each leaf, accepts a palette function.
+/// - x-label (content,none): X axis label
+/// - y-tick-step (float): Step size of y axis ticks 
+/// - x-ticks (array): List of tick values or value/label tuples
+///
+///                    *Example*
+///                    
+///                    `(1, 5, 10)` or `((1, [One]), (2, [Two]), (10, [Ten]))`
+/// - y-ticks (array): List of tick values or value/label tuples
+///
+///                    *Example*
+///                    
+///                    `(1, 5, 10)` or `((1, [One]), (2, [Two]), (10, [Ten]))`
+/// - y-unit (content,auto): Tick suffix added to each tick label
+/// - y-label (content,none): Y axis label
+/// - y-min (number,auto): Y axis minimum value
+/// - y-max (number,auto): Y axis maximum value
+#let dendrogram(data,
+                x1-key: 0,
+                x2-key: 1,
+                height-key: 2,
+                size: (auto, 1),
+                mode: "vertical",
+                line-style: (stroke: black + 1pt),
+                x-label: none,
+                y-tick-step: auto,
+                x-ticks: auto,
+                y-ticks: (),
+                y-unit: auto,
+                y-label: none,
+                y-min: auto,
+                y-max: auto,
+                ) = {
+  import draw: *
+
+  assert(mode in dendrogram-modes,
+    message: "Invalid dendrogram mode. Use: " + repr(dendrogram-modes))
+  assert(type(x1-key) in (int, str))
+  assert(type(x2-key) in (int, str))
+  assert(type(height-key) in (int, str))
+
+  if size.at(0) == auto {
+    size.at(0) = (data.len() + 2)
+  }
+
+  let max-value = (dendrogram-max-value-fn.at(mode))(data, height-key)
+  if y-max != auto {
+    max-value = y-max
+  }
+  let min-value = (dendrogram-min-value-fn.at(mode))(data, height-key)
+  if y-min != auto {
+    min-value = y-min
+  }
+
+  let x-ticks = x-ticks
+  if (x-ticks == auto) {
+    // Pre-calculate order of leaf indices
+    let x-counter = 0
+    let ticks = ()
+
+    for (idx, entry) in data.enumerate() {
+      let x1 = entry.at(x1-key)
+      let x2 = entry.at(x2-key)
+
+      // Only check relevent entries
+      if ( x1 < (data.len() + 2) ){
+        x-counter = x-counter + 1
+        ticks.push( (x-counter, x1 ))
+      }
+
+      if ( x2 < (data.len() + 2) ){
+        x-counter = x-counter + 1
+        ticks.push( (x-counter, x2 ))
+      }
+    }
+
+    x-ticks = ticks
+  }
+
+  let y-unit = y-unit
+  if y-unit == auto {
+    y-unit = []
+  }
```

Comment:
This could be simplified by allowing `y-unit` to be `none`.

Reply:
done